### PR TITLE
document CI

### PR DESCRIPTION
**Pull request by pngwn**
## Description

[Rendered](https://github.com/gradio-app/gradio/blob/update-test-strategy-document/testing-guidelines/ci.md)

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

**Comment by pngwn** [comment]
Still working on the technical section.

File: testing-guidelines/ci.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
+# Continous Integration
+
+The CI for Gradio uses GitHub Actions and almost all of the configuration to run the CI exists within the repo. 
+
+The one two cardinal rules that we have for CI are that:
+
+- CI should run on _all_ pull requests, whether those PRs are made from forks or from a branch within the repo.
+- These runs must be secure and _never_ leak any secrets, even if the run needs to have access to secrets in order to run successfully.
+
+More information onm how we achieve this can be found in the [architecture section of this document](#architecture).
```
Comment (akx) [inline testing-guidelines/ci.md:10]:
```suggestion
More information on how we achieve this can be found in the [architecture section of this document](#architecture).
```

File: testing-guidelines/ci.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
+# Continous Integration
+
+The CI for Gradio uses GitHub Actions and almost all of the configuration to run the CI exists within the repo. 
+
+The one two cardinal rules that we have for CI are that:
```
Comment (akx) [inline testing-guidelines/ci.md:5]:
```suggestion
The two cardinal rules that we have for CI are that:
```

File: testing-guidelines/ci.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+- CI should run on _all_ pull requests, whether those PRs are made from forks or from a branch within the repo.
+- These runs must be secure and _never_ leak any secrets, even if the run needs to have access to secrets in order to run successfully.
+
+More information onm how we achieve this can be found in the [architecture section of this document](#architecture).
+
+## High-level overview
+
+Broadly speaking, CI is plit into three main parts.
```
Comment (akx) [inline testing-guidelines/ci.md:14]:
```suggestion
Broadly speaking, CI is split into three main parts. 
```

File: testing-guidelines/ci.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+| JavaScript | Unit tests      | linux            | `test-js.yml`            |       |
+| n/a        | Functional      | linux            | `test-functional/yml`    |       |
+| n/a        | Visual          | linux            | `deploy+test-visual/yml` |       |
+| n/a        | Large files     | linux            | `test-hygiene.yml`       | Checks that all files are below 5 MB |
+| n/a        | Notebooks match | linux            | `test-hygiene.yml`       | Ensures that notebooks and demos are in sync |
+
+
+One important thing to note is that we split 'flaky' and 'non-flaky' python unit/ integration test out. These test are flaky because of network requests that they make, they are typically fine but anything that can cause a red check in PRs makes us less trustworthy of our Ci and confidence is the goal! The windows test are also very slow and only test a few edge cases. The flaky and windows tests are not run in every PR but are always run against the release PR to ensure everything is working as expected prior to a release. All other checks are run for every pull request, ensuring everything will work when we merge into `main`.
```
Comment (akx) [inline testing-guidelines/ci.md:61]:
```suggestion
One important thing to note is that we split 'flaky' and 'non-flaky' Python unit/integration tests out. These tests are flaky because of network requests that they make. They are typically fine, but anything that can cause a red check in PRs makes us less trustworthy of our CI and confidence is the goal! The Windows tests are also very slow and only test a few edge cases. The flaky and windows tests are not run in every PR, but are always run against the release PR to ensure everything is working as expected prior to a release. All other checks are run for every pull request, ensuring everything will work when we merge into `main`.
```

File: testing-guidelines/ci.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+#### website
+
+When a PR is created and source code has changed, a preview of the website is created.
+
+When a PR is merged into main the production version of the website is redeployed with the latest changes. 
+
+Documentation is stored by version, `main` represents the current version of the repo which may or may not match the latest release version. The process of generating documentation is roughly like this:
+
+- In Pull Requests, `main` documentation is built from the pull request branch, reflecting the latest changes in that PR (when selecting the `main` option on the docs or guides).
+- When we merge a normal Pull Request into `main` the  documentation is built from the repo, reflecting the latest changes on `main` The demo spaces are also redeployed to Hugging Face Spaces at this point (the space variant with the `main_` prefix).
```
Comment (akx) [inline testing-guidelines/ci.md:83]:
```suggestion
- When we merge a normal pull request into `main` the documentation is built from the repo, reflecting the latest changes on `main`. The demo spaces are also redeployed to Hugging Face Spaces at this point (the space variant with the `main_` prefix).
```

File: testing-guidelines/ci.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+- When a new version of gradio is released (when a versioning PR is merged), the current documentation in the repo is deployed under a version tag. So for version `3.1.1` the current docs and guides in main will be available under that version for eternity. At this point `main` (built from source) and `3.1.1` (built from source and stored in the cloud) are equivalent. We also redeploy demo spaces when a new Gradio version is released, this time without the `main_` prefix).
```
Comment (akx) [inline testing-guidelines/ci.md:84]:
```suggestion
- When a new version of Gradio is released (when a versioning PR is merged), the current documentation in the repo is deployed under a version tag. So for version `3.1.1` the current docs and guides in main will be available under that version for eternity. At this point `main` (built from source) and `3.1.1` (built from source and stored in the cloud) are equivalent. We also redeploy demo spaces when a new Gradio version is released, this time without the `main_` prefix.
```

File: testing-guidelines/ci.md [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+
+When the `changeset-release/main` branch is merged into main all of the necessary versions will have been bumped and the changelogs generated. We then go through to see what needs to be published and publish to PyPi for Python packages and npm for javascript pacakges.
+
+
+## Architecture
+
+The CI on this repo is a little unconventional, this is mainly to work around various gaps in the GitHub Actions API while solving for our use case.
+
+The [technical details](#technical-details) below explain how things are setup, the [context](#context) sections explains what problem we are trying to solve with this architecture,why  we took this approach and the various challenges we faced along the way. It is a little 'extra' for repo docs but serves as good context for the future.
```
Comment (abidlabs) [inline testing-guidelines/ci.md:167]:
```suggestion
The [technical details](#technical-details) below explain how things are setup, the [context](#context) sections explains what problem we are trying to solve with this architecture, why we took this approach, and the various challenges we faced along the way. It is a little 'extra' for repo docs but serves as good context for the future.
```

**Comment by abidlabs** [comment]
Wow @pngwn this is incredible! Very comprehensive, thanks for writing this down. 

My one suggestion is since the "Technical Details" section is so detailed, it might be helpful to have a diagram showing the sequence of workflows that are run when files are changed (you had shared something like this internally) at the beginning of the section to orient readers.

**Review by abidlabs — APPROVED** [review]
🔥

**Comment by pngwn** [comment]
@abidlabs Yeah, that's a great idea. I'll try to come up with a nice diagram to show the flow.

**Comment by akx** [comment]
> I'll try to come up with a nice diagram to show the flow.

You [can include inline](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/) [Mermaid](https://mermaid.live/) diagrams in GitHub Markdown, if that helps!

```mermaid
flowchart TD
    A[wow] -->|so| B[cool]
```

File: testing-guidelines/ci.md [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+- Versioning and Publishing
+
+### When do checks run
+
+Checks only run when needed but are required to pass when they run.
+
+We check to see which source files have changed and run the necessary checks. A full breakdown of how we determine this for each kind of check can be found in the [`changes` action](https://github.com/gradio-app/gradio/blob/main/.github/actions/changes/action.yml#L65-L108) but the high-level breakdown is as follows:
+
+- __python checks__ - whenever python source, dependencies or config change.
+- __javascript checks__ - whenever javascript source, dependencies or config change.
```
Comment (akx) [inline testing-guidelines/ci.md:27]: Suggesting to unify the spelling of Python and JavaScript in this doc.

```suggestion
- __Python checks__ - whenever Python source, dependencies or config change.
- __Javascript checks__ - whenever JavaScript source, dependencies or config change.
```

File: testing-guidelines/ci.md [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+| JavaScript | Unit tests      | linux            | `test-js.yml`            |       |
+| n/a        | Functional      | linux            | `test-functional/yml`    |       |
+| n/a        | Visual          | linux            | `deploy+test-visual/yml` |       |
+| n/a        | Large files     | linux            | `test-hygiene.yml`       | Checks that all files are below 5 MB |
+| n/a        | Notebooks match | linux            | `test-hygiene.yml`       | Ensures that notebooks and demos are in sync |
+
+
+One important thing to note is that we split 'flaky' and 'non-flaky' Python unit/integration tests out. These tests are flaky because of network requests that they make. They are typically fine, but anything that can cause a red check in PRs makes us less trustworthy of our CI and confidence is the goal! The Windows tests are also very slow and only test a few edge cases. The flaky and windows tests are not run in every PR, but are always run against the release PR to ensure everything is working as expected prior to a release. All other checks are run for every pull request, ensuring everything will work when we merge into `main`.
```
Comment (akx) [inline testing-guidelines/ci.md:61]: Splitting to multiple lines by sentence so future diffs are easier, fix Windows case.
```suggestion
One important thing to note is that we split 'flaky' and 'non-flaky' Python unit/integration tests out.
These tests are flaky because of network requests that they make. They are typically fine, but anything that can cause a red check in PRs makes us less trustworthy of our CI and confidence is the goal!
The Windows tests are also very slow and only test a few edge cases.
The flaky and Windows tests are not run in every PR, but are always run against the release PR to ensure everything is working as expected prior to a release.
All other checks are run for every pull request, ensuring everything will work when we merge into `main`.
```

File: testing-guidelines/ci.md [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+Versioning is made up of two components:
+
+- Each pull request has to explain what kind of change it introduces.
+- When a pull request is merged into `main` we need to figure out what the next version for package should be based on the combination of all changes and whether or not we should publish.
+
+To manage versioning and publishing we use a combination of the [`changesets`](https://github.com/changesets/changesets) library and some custom additions.
+
+#### pull request changesets
+
+Each pull request must include a 'changeset', this is a simple markdown file that states what _type_ of change it is, what version bump is required for each affected package, and a description of that change.
```
Comment (akx) [inline testing-guidelines/ci.md:127]:
```suggestion
Each pull request must include a changeset file. This is a simple markdown file that states what _type_ of change it is, what (if any) version bump is required for each affected package, and a description of that change.
```

File: testing-guidelines/ci.md [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+
+Change types are purely semantic and we only use them to organise the changelog, these 'types' are related to bump types but they are not the same. They can be one of the following:
```
Comment (akx) [inline testing-guidelines/ci.md:129]:
```suggestion
Change types are purely semantic and we only use them to organise the changelog. These "types" are related to bump types but they are not the same. They can be one of the following:
```

File: testing-guidelines/ci.md [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+
+### Technical details
+
+Most our our workflows are contained within the repository but we also have some custom JavaScript actions that are stored in the [gradio-app/github](https://github.com/gradio-app/github) repository. These actions will be mentioned below but I'll put a list here for ease of navigation:
+
+- [`find-pr`](https://github.com/gradio-app/github/tree/main/packages/find-pr) - for getting a pr number, pr branch, and other metadata for a `workflow_run`.
+- [`comment-pr`](https://github.com/gradio-app/github/tree/main/packages/comment-pr) - for creating and updating comments on a pull request.
+- [`commit-status`](https://github.com/gradio-app/github/tree/main/packages/commit-status) - for creating statuses on commits.
+- [`generate-changeset`](https://github.com/gradio-app/github/tree/main/packages/generate-changeset) - for generating and updating changesets based on the state or a pr and branch.
+- [`publish-pypi`](https://github.com/gradio-app/github/tree/main/packages/publish-pypi) - for publishing packages to pypi.
```
Comment (akx) [inline testing-guidelines/ci.md:177]: I've had great success using pypa's own action for the actual publishing, no need to faff around with twine and secrets and so on (especially when using trusted publishers): https://github.com/marketplace/actions/pypi-publish

File: testing-guidelines/ci.md [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+
+### Technical details
+
+Most our our workflows are contained within the repository but we also have some custom JavaScript actions that are stored in the [gradio-app/github](https://github.com/gradio-app/github) repository. These actions will be mentioned below but I'll put a list here for ease of navigation:
+
+- [`find-pr`](https://github.com/gradio-app/github/tree/main/packages/find-pr) - for getting a pr number, pr branch, and other metadata for a `workflow_run`.
+- [`comment-pr`](https://github.com/gradio-app/github/tree/main/packages/comment-pr) - for creating and updating comments on a pull request.
+- [`commit-status`](https://github.com/gradio-app/github/tree/main/packages/commit-status) - for creating statuses on commits.
+- [`generate-changeset`](https://github.com/gradio-app/github/tree/main/packages/generate-changeset) - for generating and updating changesets based on the state or a pr and branch.
```
Comment (akx) [inline testing-guidelines/ci.md:176]: TBH, I've been wondering (for my own projects and so on, but the same wonderment applies here) why not write this stuff into a couple Python scripts instead...

File: testing-guidelines/ci.md [hunk 14]
```diff
@@ -0,0 +1,431 @@
...
+- `spaces` - previews a series of gradio apps on Huggin Face Spaces as they will appear and behave when the pull requests is merged and released.
+
+The `storybook` and `website` previews are relatively straightforward as they use tooling provided by the service providers that we use in order to deploy. We just make sure each run has the necessary assets and components built.
+
+The `spaces` preview is a little more involved as it is a custom process and requires us to build a custom gradio wheel from that pull request.
+
+The process is relatively straightforward, and follows [the steps mentioned above](#spaces) but there a few details to be aware of.
+
+- We use [a custom script](https://github.com/gradio-app/gradio/blob/main/scripts/copy_demos.py) to pull in a select number of spaces and build them into a single FastAPI application. We serve each demo on its own subpath. This is the demo app that gets deployed to spaces.
+- We build a new wheel from the pull requests source code and upload it to s3, we then add the url for this wheel to the requirements.txt of the space we are deploying.
```
Comment (akx) [inline testing-guidelines/ci.md:233]: FWIW, https://github.com/aws-actions/configure-aws-credentials supports GitHub's OIDC too these days, so no need for secret tokens...

File: testing-guidelines/ci.md [hunk 15]
```diff
@@ -0,0 +1,431 @@
...
+- We want CI to be fast, certain jobs should be skipped where appropriate. These jobs may still be 'required' to pass if—and only if—they run.
+- CI should give confidence over time, we don't want to be overlooking the odd ❌ as a 'normal' part of our CI, even for optional jobs. This will erode trust.
+- Many of our CI jobs share the same steps. We want to avoid excessive duplication where possible for maintenance reasons.
+
+Some of these are discrete problems with their own discrete solutions but a lot of the challenges stem from when certain GitHub Action events occur and how priveleged/ secure those event 'types' are.
+
+#### Demystifying event triggers
+
+Workflows are a discreet set of jobs with a discrete set of steps. It might be reasonable to assume that a workflow is a workflow. Sadly this isn't true, the event that triggers the workflow dictates not only when that workflow will run (which makes sense) but also a bunch of other information about both its environment and even which version of that workflow file will run (this is a git repo after all). This latter feature, _also_ makes sense, but it isn't immediately apparent.
```
Comment (akx) [inline testing-guidelines/ci.md:332]:
```suggestion
Workflows are a discrete set of jobs with a discrete set of steps. It might be reasonable to assume that a workflow is a workflow. Sadly this isn't true, the event that triggers the workflow dictates not only when that workflow will run (which makes sense) but also a bunch of other information about both its environment and even which version of that workflow file will run (this is a git repo after all). This latter feature _also_ makes sense, but it isn't immediately apparent.
```

File: testing-guidelines/ci.md [hunk 16]
```diff
@@ -0,0 +1,431 @@
...
+
+- `pull_request` - This event runs correctly on contributor PRs and check out the correct branch by default (more on this later) but it doesn't not have access to secrets.
+- `pull_request_target` - Same as `pull_request` but it _does_ have access to secrets. However because this event runs in to context of the branch (and repo) the PR is made from, that PR has also has direct access to secrets making it insecure.
+- `push` - When triggered from a fork, this will essentially be a `push` to the fork not the target repo. In many cases the workflow won't run ata ll (depends on the settings of the fork) and it won't have access to secrets from the base repo. Even if it did, it would be inscure.
```
Comment (akx) [inline testing-guidelines/ci.md:336]:
```suggestion
- `push` - When triggered from a fork, this will essentially be a `push` to the fork, not the target repo. In many cases the workflow won't run at all (depends on the settings of the fork) and it won't have access to secrets from the base repo. Even if it did, it would be insecure.
```

File: testing-guidelines/ci.md [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+
+There are ways to run workflows indirectly:
+
+- `workflow_dispatch` - This event always runs in the context of main  You can programatically trigger this workflow event, allowing more control over where that workflow runs but you need to use the GitHub API to do this. Ttherefore the triggering workflow needs access to secrets, rendering it insecure for our purposes.
```
Comment (akx) [inline testing-guidelines/ci.md:340]:
```suggestion
- `workflow_dispatch` - This event always runs in the context of `main`. You can programatically trigger this workflow event, allowing more control over where that workflow runs but you need to use the GitHub API to do this. Therefore the triggering workflow needs access to secrets, rendering it insecure for our purposes.
```

File: testing-guidelines/ci.md [hunk 18]
```diff
@@ -0,0 +1,431 @@
...
+  <summary>What does the "context" of an event mean?</summary>
+
+In GitHub Actions 'context' is a somewhat overloaded term, but it typically describes the permissions, available data, and the source code state that a given workflow has access to, without any additional code. 
+
+For example, you can check out any branch of any public repo in any workflow but the context is important before any configured steps are run, in fact, the context is important before the workflow even starts. For practical purposes, there are two elements to the 'context' that people care about.
+
+- Which workflow file on which branch actually runs.
+- What information about the repo or triggering event does that workflow have access to.
+
+If a workflow "runs in the context of the default branch" then it will use the workflow that exists on trhe default branch, regardless of whether or not the event that originall triggered it was on another branch. If the workflow "runs in the context of the pull request branch" then it will pull the workflow file from the pull request branch.
```
Comment (akx) [inline testing-guidelines/ci.md:353]:
```suggestion
If a workflow "runs in the context of the default branch" then it will use the workflow that exists on the default branch, regardless of whether or not the event that originally triggered it was on another branch. If the workflow "runs in the context of the pull request branch" then it will pull the workflow file from the pull request branch.
```

File: testing-guidelines/ci.md [hunk 19]
```diff
@@ -0,0 +1,431 @@
...
+
+The information available inside a workflow after it has started (usually available via the [`github` context](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context)). For pull requests, this will include things like the pull request number, and the ref and HEAD SHA of the pull request branch. For workflows running in the context of the default branch, this may not contain much information, but all references to the branch and sha will mainly just be references to main.
+
+</details>
+
+#### New solution, new problems
+
+For the reasons described above, we chose to use `workflow_run` _heavily_ for the gradio repo. However `workflow_run` presents its own challenges:
+
+- This event runs in the context of main, it doesn't offer any of the conveniences that `push` and `pull_request` events give you, it knows very very little about the workflow run even that triggered it. It _does not_ inherit the teriggering workflow's context. This is a huge problem.
```
Comment (akx) [inline testing-guidelines/ci.md:363]:
```suggestion
- This event runs in the context of main, it doesn't offer any of the conveniences that `push` and `pull_request` events give you, it knows very very little about the workflow run even that triggered it. It _does not_ inherit the triggering workflow's context. This is a huge problem.
```

File: testing-guidelines/ci.md [hunk 20]
```diff
@@ -0,0 +1,431 @@
...
+- This workflow kind of runs in the void. It is run in the context of the default branch and so maintains references to that branch, however, it isn't really 'attached' to a commit or ref in any meaningful way and the status of the run (the 'check') is not added to any commits anywhere.
+
+Both of these problems were eventually solve by using the GitHub API in combination with the information we get from the workflow event's context. Getting the commit reference of the pull request that triggered the workflow is the main challenge, when we have that, creating statuses on commits is trivial. 
+
+##### What branch am i even in?
```
Comment (akx) [inline testing-guidelines/ci.md:368]:
```suggestion
##### What branch am I even in?
```

File: testing-guidelines/ci.md [hunk 21]
```diff
@@ -0,0 +1,431 @@
...
+
+The 'context' a workflow runs in is the branch that actions/checkout will checkout by default. In this case that is `main`. We don't want `main`.
+
+Figuring out what branch or pull request triggered a workflow run is surprisingly difficult depending on the event that you allow to trigger it. The data you have access to in a ` workflow_run` event is pretty limited. It is okay for pull requests, you get the PR number, but for pushes and other events it can be challenging. We trigger `workflow_runs` from the following events:
```
Comment (akx) [inline testing-guidelines/ci.md:372]:
```suggestion
Figuring out what branch or pull request triggered a workflow run is surprisingly difficult depending on the event that you allow to trigger it. The data you have access to in a `workflow_run` event is pretty limited. It is okay for pull requests, you get the PR number, but for pushes and other events it can be challenging. We trigger `workflow_run`s from the following events:
```

File: testing-guidelines/ci.md [hunk 22]
```diff
@@ -0,0 +1,431 @@
...
+
+- `pull_request` - this is fine, we get the PR number.
+- `push` - not fine, but we get the commit SHA which will do.
+- `issue_comment` - couldn't be less fine, we only get the title.
+
+*The reason we use the `issue_comment` event is because pull request are actually a special type of issue when it comes to the title, body, and replies.*
```
Comment (akx) [inline testing-guidelines/ci.md:378]: (GitHub likes this pattern. Similarly, organizations on GitHub are a special kind of user.)

File: testing-guidelines/ci.md [hunk 23]
```diff
@@ -0,0 +1,431 @@
...
+- Any labels for the PR (we use these for certain conditional jobs)
+
+<details>
+<summary>A magical branch, you say?</summary>
+
+GitHub actually creates two magical refs. `pull/<pr-number>/head` and `pull/<pr-number>/merge`. Both of these refs are read-only, you cannot push to them no matter how many `-f`s you add.
+
+The `head` variant is pretty much the same as the HEAD of the pr branch, except it exists in the target repo regardless of whether it was created from a fork or not. This is quite nice as the commit SHA for this ref will be the same as the commit SHA for the HEAD of the source branch. This makes checking out the branch easier.
+
+The `merge` variant is special. This is a branch that merges the PR changes into the target branch. `pull_request` events have this branch set as their 'default' and it is what gets checked out by default in `pull_request` workflows. The beauty of this branch is that any tests you run against it are essentially being run on the merged result of this PR and `main`. This isn't commonly know but it is exactly what you want in a pull request.
```
Comment (akx) [inline testing-guidelines/ci.md:396]:
```suggestion
The `merge` variant is special. This is a ref that has merged the PR changes into the target branch. `pull_request` events have this ref set as their 'default' and it is what gets checked out by default in `pull_request` workflows. The beauty of this ref is that any tests you run against it are essentially being run on the merged result of this PR and `main`. This isn't commonly known but it is exactly what you want in a pull request.
```

File: testing-guidelines/ci.md [hunk 24]
```diff
@@ -0,0 +1,431 @@
...
+This sounds contradictory but what we want is check that don't _always_ need to run but when they run they _must_ pass. GitHub doesn't really have a concept of this.
+
+The solution is to set the check as required in the repo settings and then do the following:
+
+- If the job runs then the commit status is set to pending prior to the run.
+- If the job fails then the commit status should be set to failed.
+- If the job successed then then commit status should be set to success.
+- If the job does not need to run then it should be set to success with some text explaining it was skipped.
+
+Determining what has changed is straightforward, we use a third-party action for this. But we clear need to run but also maybe not run our workflows.
```
Comment (akx) [inline testing-guidelines/ci.md:416]: > But we clear need to run but also maybe not run our workflows.

whuh?

File: testing-guidelines/ci.md [hunk 25]
```diff
@@ -0,0 +1,431 @@
...
+
+To solve this particular problem we _always_ trigger our workflows but don't always run all of them.
+
+- Every workflow we might want to run is triggered by the pull request. We have a simple workflow that does nothing, it simply acts as a 'hook' for the `workflow_run` workflows to listen to.
+- Those workflows have their own information about whether the job should run or not.
+- If the job thinks that it _should_ run then it creates a 'pending' status and sets its output to `should_run = true`
+- If the job thinkgs that it _shouldn't_ run then it creates a 'success' status nand sets its output to `should_run = false`. 
+- The next job in the workflow _depends_ on that initial run. It will only run on the condition that the `changes` job has an output of `should_run == true`. 
+- If it does run then the workflow does its thing and then updates the commit status to `success` or `failure` depending on the outcome.
```
Comment (akx) [inline testing-guidelines/ci.md:425]:
```suggestion
- If it does run, the workflow does its thing and then updates the commit status to `success` or `failure` depending on the outcome. 
```

File: testing-guidelines/ci.md (hunk 12 above)
Comment (pngwn) [reply testing-guidelines/ci.md:177]: We used to use it but its really slow because it is a docker action, so i wrote a js one that is much faster.

File: testing-guidelines/ci.md (hunk 24 above)
Comment (pngwn) [reply testing-guidelines/ci.md:416]: I think my brain broke

File: testing-guidelines/ci.md (hunk 12 above)
Comment (akx) [reply testing-guidelines/ci.md:177]: It isn't really slow at all for me – it does depend on bits being cached of course...

Comment (pngwn) [reply testing-guidelines/ci.md:177]: Its spins up a docker container, so its always going to be slower than a js action but each to their own.

File: testing-guidelines/ci.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (pngwn) [reply testing-guidelines/ci.md:396]: I'm pretty sure it is but it might not always be true. I'd need to check but it is possible that `pull/<pr>/merge` is preferred but it falls back to `pull/<pr>/head`. They would know ahead of time if it is possible to run checks on the former because it is created every time a new commit is pushed prior to running checks.

If there are no conflicts, it definitely uses the merged result though, that I have tested.

File: testing-guidelines/ci.md (hunk 12 above)
Comment (pngwn) [reply testing-guidelines/ci.md:181]: Oh the other thing is you have to use the action once per publish, so if you are publishing multiple packages, it isn't ideal.

File: testing-guidelines/ci.md (hunk 13 above)
Comment (pngwn) [reply testing-guidelines/ci.md:180]: I guess it depends, we publish a lot of packages and want the process to be very simple, this allows us to write less code ourselves. There is no reason why a few scripts couldn't replace this for a specific case.

That said, I do have vague plans to pull this out and package it up in way that requires less custom code and just some simple config.

**Comment by pngwn** [comment]
Thanks for the reviews folks, I went for a static image diagram for now because I lost a lot of time trying to make mermaid look not terrible. I'll revisit that at some point, it would be nice to have the flow diagram in the repo for easier maintenance.

**Comment by pngwn** [comment]
ty @abidlabs and @akx i appreciate the thorough reviews!
